MoveProvisionedDashboard: bug fix when selecting root folder, error is showing

ywzheng1 opened the pull request:
**What is this feature?**

- Fix when user moving resources, selecting root folder is causing error.  
- Bugfix when whole instance is provisioned, root folder selection is behave incorrectly 

Multiple provisioned folder:  

https://github.com/user-attachments/assets/98faaf6f-7f72-40c9-8810-ac15494f0827




**Why do we need this feature?**

- User can move resources around correctly. 

**Who is this feature for?**

git users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/git-ui-sync-project/issues/407
Fixes https://github.com/grafana/git-ui-sync-project/issues/452
Fixes https://github.com/grafana/git-ui-sync-project/issues/350

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

ywzheng1 committed:
MoveProvisionedDashboard: root folder error fix

ywzheng1 committed:
add unit tests

ywzheng1 committed:
comment

ywzheng1 commented on public/app/features/browse-dashboards/state/reducers.ts:
```diff
@@ -184,8 +184,8 @@ export function setAllSelection(
         }
 
-        // Skip all provisioned resources during "select all" on root level
-        if (child.managedBy === ManagerKind.Repo && !child.parentUID) {
```
switched to use excludeUIDs which is more manageable. where parents controls business logic

MissingRoberto reviewed (dismissed):
See the issue described above. 

The same issues with the badge happens in the folder selection dropdown: 

<img width="441" height="347" alt="image" src="https://github.com/user-attachments/assets/223f6c96-c549-4c39-8561-2f9b8050fade" />

ywzheng1 committed:
Merge remote-tracking branch 'origin' into ywzheng1/moving-resource-to-repo-top-level-error

ywzheng1 committed:
remove provisioned badge when whole instance is provisioned, root folder checkbox fix

ywzheng1 committed:
lint

Clarity-89 commented on public/app/features/browse-dashboards/components/BulkActions/utils.ts:
```diff
@@ -16,12 +16,64 @@ export interface BulkActionProvisionResourceProps {
+  targetFolder,
+  repoName,
+  hidePrependSlash = false,
+}: {
+  targetFolderUID?: string;
+  targetFolder?: Folder;
+  repoName?: string;
+  hidePrependSlash?: boolean;
+}): string | undefined {
```
Nit: can we move the type out?

ywzheng1 replied:
fixed

Clarity-89 commented on public/app/features/browse-dashboards/components/BulkActions/utils.ts:
```diff
@@ -16,12 +16,64 @@ export interface BulkActionProvisionResourceProps {
+/**
+ * Checks if the target folder is the repository root folder
+ */
+function isRepositoryRootFolder(targetFolder?: Folder, repoName?: string): boolean {
```
```suggestion
function isRepositoryRootFolder(targetFolder?: Folder, repoName?: string) {
```
By relying on inferred type returns (for simple types) we can make the code a bit easier to maintain.

ywzheng1 committed:
Update public/app/features/browse-dashboards/components/BulkActions/utils.ts

Co-authored-by: Alex Khomenko <Clarity-89@users.noreply.github.com>

ywzheng1 committed:
type fix

ywzheng1 commented:
@Clarity-89 Noted! This PR includes 3 linked tickets since the bug tickets were created after I started it, so I linked them together 😅 will make sure to separate them carefully in the future

ywzheng1 committed:
empty repo folder selection error fix

ywzheng1 commented:
@Clarity-89 ah good catch, because we never reset the `hasSubmitted` state to false after error. Fixed this. Also moved error from console error to form validation error so its more clear to user

<img width="732" height="455" alt="image" src="https://github.com/user-attachments/assets/82ceca52-5785-4bb5-b4ee-9c23f4e99e9c" />

Clarity-89 reviewed (commented):
Noticed that select all checkbox doesn't select the items anymore, not sure if it was caused by this PR:

<img width="552" height="621" alt="Screenshot 2025-08-20 at 8 45 31" src="https://github.com/user-attachments/assets/01813e37-6e09-499b-87ea-a650ae264c9c" />

ywzheng1 commented:
@Clarity-89 we disabled select all on provisioned root provisioned folder (if its only one folder provisioned). Conversation here: https://raintank-corp.slack.com/archives/C08T7NVBCUA/p1754399579432199

Clarity-89 commented:
> @Clarity-89 we disabled select all on provisioned root provisioned folder (if its only one folder provisioned). Conversation here: https://raintank-corp.slack.com/archives/C08T7NVBCUA/p1754399579432199
> 

Got it, but it does look weird when you have only one provisioned folder and select all button does nothing. Should we disable pr hide that checkbox when there's only provisioned items at the root? 

> Interesting that on my end, we can still select subitems tho: 

That doesn't look like subitems but separate non-provisioned folders?

ywzheng1 commented on public/app/features/browse-dashboards/state/reducers.ts:
```diff
@@ -97,11 +96,6 @@ export function setItemSelectionState(
   }
 
-  // Prevent selection of root provisioned folders
-  if (item.managedBy === ManagerKind.Repo && !item.parentUID) {
```
The previous selection logic was becoming increasingly complex and scattered across multiple validation hooks (useSelectionRepoValidation, useSelectionProvisioningStatus, etc.) as we added support for provisioning folders.

So we've centralized the exclusion logic using excludeUIDs in BrowseView.tsx [here](https://github.com/grafana/grafana/blob/ywzheng1/moving-resource-to-repo-top-level-error/public/app/features/browse-dashboards/components/BrowseView.tsx#L45-L56). This gets passed to the selection reducer where it's handled with a simple check:
```
// Skip items in the exclude list
if (excludeUIDs?.includes(child.uid)) {
  continue;
}
```
This approach scales much better as we add more folder types and selection rules in the future. (considering team folders)

ywzheng1 commented:
@Clarity-89 yep they are local 😅 Provisioned folder subitems are not selectable on root level. 
<img width="747" height="451" alt="image" src="https://github.com/user-attachments/assets/def07ebc-c5d5-4e7b-8677-a8ac6a5a33db" />

Yep we should do that! If only one provisioned folder, treat it as whole instance is provisioned.

ywzheng1 committed:
If root level only have one provisioned folder, allow select all on subitems

ywzheng1 committed:
remove button tooltip and remove comment

ywzheng1 committed:
regenerate API clients after schema updates

ywzheng1 committed:
add comment

ywzheng1 commented on public/app/features/browse-dashboards/components/BulkActions/utils.ts:
```diff
@@ -16,12 +17,67 @@ export interface BulkActionProvisionResourceProps {
+  hidePrependSlash = false,
+}: GetTargetFolderPathInRepoParams): string | undefined {
+  const ROOT_PATH = '/';
+  const EMPTY_PATH = '';
```
Good point about the empty string convention being unclear. I've added a clarifying comment to make the intent more explicit. I'll explore cleaner alternatives in future PRs to improve readability.

Clarity-89 reviewed (approved):
You'll need to re-generate the API clients for the CI tests to pass.

ywzheng1 committed:
generate api clients
